### ISSUE TITLE
Futures support

### ISSUE THREAD

**Issue by NoraCodes** [zesterer/flume#2]
I'm currently looking at what an API might look like for `async` support. At the most basic, I'd want a `async fn on_recv` and `async fn on_send`, but it would also be nice to have a method producing an `impl async_std::stream::Stream`.

Is this in line with the project's goals?

**Comment by zesterer** [comment]
If we can keep the implementation safe and relatively simple, I'd be very happy to include this. I'll try to find some time to look into the mechanics of how this is likely to work. If this is something you're willing to chase, then I'm happy to provide what support I can.

**Comment by najamelan** [comment]
I would also be interested. Most useful is `Stream` and `Sink` impl. In practice, what it comes down to is:

- you shouldn't ever block the thread, which means need for different sync primitives I'm afraid
- when no data is available on receiver or the buffer is full on sender, you must return Pending and save the waker to be woken up by the other side when the condition changes. eg. when sending to an empty buffer, you must check if the reader was pending and wake up the waker. when reading from a full buffer, you must check if any writers are pending and wake one them up. This is a bit more complicated as waking up all for a single free slot might cause thundering horde problem and not waking up enough when several slots are available will not use the buffer optimally. A further problem is that a woken up task is not always guaranteed to actually use it's slot, in which case you must wake another one up... 

maybe it's more useful to have an async-flume crate because it does change the internals quite a bit I think, and it can add overhead on the sync version. I think the ring_channel crate did something like this though. I haven't reviewed the final version to see how well it succeeded, but it might provide inspiration for combining both sync and async in the same crate.

**Comment by Restioson** [comment]
Just btw, I'm playing around with this in a fork atm.

**Comment by Restioson** [comment]
Extremely early impl, probably nonfunctional is here https://github.com/Restioson/flume/tree/async

**Comment by najamelan** [comment]
@Restioson It would have been good to commit the file before making changes. Makes it much easier to follow along.

**Comment by Restioson** [comment]
@NoraCodes
> At the most basic, I'd want a async fn on_recv and async fn on_send

As far as I can tell, the latency introduced by yielding to the executor *far* outstrips the cost of just spinning waiting for the channel's queue to be ready. From some preliminary testing, it takes about 10us with yielding to the reactor and 120ns without (81x slower) in a workload with one reader which is receiving in a loop and one writer which is writing in a loop (in parallel). So, I'm not sure if this interface would be useful.

**Comment by Restioson** [comment]
#7 is the WIP PR for this

**Comment by zesterer** [comment]
Simple asynchronous receiving support has been added in `0.7.0`, just released. Hopefully we can extend this to support bounded queue sending, `Sink` and `Stream` soon.

**Comment by Restioson** [comment]
#11 adds a stream impl

**Comment by Restioson** [comment]
@najamelan

> you shouldn't ever block the thread

So far, there is a singe lock used to wake up any waiting senders. I'm not sure how long this locks for or if/how we can get rid of it.

> maybe it's more useful to have an async-flume crate because it does change the internals quite a bit I think

This is being looked at but so far it doesn't change *that* much.

> it can add overhead on the sync version

True. This is why it's feature-gated and can be disabled.

**Comment by najamelan** [comment]
Thanks for working on this. I will try to find time to review this, but don't wait for me...

**Comment by Restioson** [comment]
#11 landed so there is now a stream impl (also FusedFuture and FusedStream)

**Comment by Restioson** [comment]
#20 adds Sink and async_send methods

**Comment by najamelan** [comment]
I had time to have a first look over the code. I haven't yet tried using it, so these are first impressions:

### SenderSink type
This type is detached from the actual `Sender`. I think the design has a number of issues:
- it now has a lifetime. I have not tried using it, but I expect that this might make it difficult to use with API's/combinators that take a Sink by value, as the Sink is valid for a limited lifetime.
- it has it's own buffer, always ready to accept new messages, effectively turning bounded channels into unbounded ones and not providing back pressure. I would have expected there to be at most one item in a sink specific buffer, something like `Option<T>`. Then if there is already one, `poll_ready` returns `Pending`.
- not sure how useful it is to check for disconnected again in `start_send`.
- does not implement clone. For API's wanting to use a mpsc type channel, I would expect them to want a type that is `T: Sink + Clone`.
- poll_flush will push messages back into the sink's buffer when disconnected, but how does it proceed from there? I imagine a disconnected channel never becomes operational again, and there is no way for the user to recover the messages in that buffer, so why is it kept?

### Wakers
- receivers will wake one Sender. This avoids the thundering herd problem, but can deadlock if that sender doesn't actually send any message. The others won't necessarily be woken up. It's kind of hard to deadlock as long as there are items remaining in the buffer and readers keep reading, but I think it needs to be properly evaluated and tested/documented. Some prior discussion about this problem can be [found here](https://github.com/rust-lang/futures-rs/issues/800). I haven't dived in the solution supposedly [in this PR](https://github.com/rust-lang/futures-rs/pull/985).

### Nits
- `Receiver::poll` has a redundant else block near the top that can be eliminated, since the if block returns from the function anyways.
- waking oneself when it's not possible to take a lock can work, but it can also keep looping, consuming CPU. I'm not sure how much difference it makes in what scenario, but a futures aware lock that can be awaited might be more appropriate. That's harder to graft on a synchronous lib though. I wonder if it's not cleaner to fork flume and make an async version that is completely tailored to async, as it would probably keep code simpler and I don't expect people to combine sync and async usage.

**Comment by Restioson** [comment]
> This type is detached from the actual Sender

It has to be, unless all senders have a buffer for messages that need to be sent, by nature of how the Sink trait works. Could make an owned version too but I went for a borrowing one initially so as to avoid a clone cost.

> it has it's own buffer, always ready to accept new messages

It has to, because of `start_send`. There is no way to check the current available capacity in `start_send` without doing as much work as actually sending it - which requires blocking. 

>  there is no way for the user to recover the messages in that buffer

There is: `into_pending`. I think I meant to push as many as the queue could fit in `poll_flush` though, which I need to fix.

>  I wonder if it's not cleaner to fork flume and make an async version that is completely tailored to async

It *might* be, and that was originally how it worked, but we decided to put them together because we thought it'd be simpler to use.

> A futures aware lock that can be awaited might be more appropriate

It would, but like you I'm not sure how this would integrate with the sync api.

**Comment by najamelan** [comment]
Just a quick reaction for now.

> It has to, because of start_send. There is no way to check the current available capacity in start_send without doing as much work as actually sending it - which requires blocking.

I understand what you mean, but I think just 1 slot is enough and `poll_ready` should return pending if the slot is already taken, which would not let it accept an unbound amount of messages. It would then become bounded + 1 per sender sink. The contract is that users should never call `start_send` unless `poll_ready` gives green light.

I think the reasoning behind this API is that you can check if you can send without passing in the message, so in the case of failure you don't need to give the message back to the client to try later. That's why `start_send` doesn't return `Poll`.

So to go completely without an extra buffer, `poll_ready` would have to somehow reserve the slot, but that's got a bunch of synchronization problems of it's own, like what if `start_send` never get's called, for how long you reserve etc. 

Or it would have to keep a lock until `start_send` is called so nobody else can interfere until that has happened...

**Comment by Restioson** [comment]
> I understand what you mean, but I think just 1 slot is enough and poll_ready should return pending if the slot is already taken

Good idea! Maybe this could exist as well as some kind of buffered sink (what we already have).

> you can check if you can send without passing in the message

Yep. The issue here is that the amount of work needed to check if you can send is basically the same as the amount of work to send on its own, and would require blocking (need to lock tthe inner queue).

**Comment by najamelan** [comment]
> Maybe this could exist as well as some kind of buffered sink (what we already have).

Yes, of course. But why not put the field on `Sender` and feature gate it. This way Sink can directly be implemented for `Sender`.

**Comment by Restioson** [comment]
It's very different behaviour to the other approach though, doesn't feel right for a feature flag. We could just hold an Option<T> in the sender to allow unbuffered sink

**Comment by najamelan** [comment]
Well there is several other feature gated fields already.

**Comment by najamelan** [comment]
I was doing some research on this topic and found two discussions on the futures repo that where enlightening: rust-lang/futures-rs#1312 and rust-lang/futures-rs#984.

**Comment by najamelan** [comment]
Some more discussion happening around Sink: rust-lang/futures-rs#2109. Also I think there is a deadlock in the current implementation:

When any reader or writer obtains the lock, now all others will cannot make progress, nor store their waker, as the lock is occupied. When the lock is released, there are no wakers to be waken up. So only this user can still do something, because all the others are waiting to be woken up, if now this cannot make progress, either because it's a reader and the queu is empty, or it's a writer and the queue is full, this one will also store it's waker and a deadlock happens.

Note that `try_recv` does not wake up any wakers when the queue is empty.  Neither does `Sender::poll`. I quickly tried to solve these last two, and I'm still running into deadlocks, so proper multithreaded testing really is in order here.

**Comment by zesterer** [comment]
@najamelan Have you taken a look at the implementation on the `flavor-refactor` branch? I think it's likely going to become the new implementation of this lib (it's much faster in most benchmarks and solves a variety of other issues). It doesn't yet have support for `select` of `async` though.

**Comment by najamelan** [comment]
@zesterer, no I have only reviewed the async implementation that is merged into master. I haven't reviewed anything synchronous.

**Comment by zesterer** [comment]
Fulled addressed by #24

**Comment by Restioson** [comment]
@najamelan if you still want to review the latest on async, there's PR #40 which contains a few tweaks
